### PR TITLE
max_version support

### PR DESCRIPTION
**Pull request by wangnengjie**
Signed-off-by: wangnengjie <751614701@qq.com>

Support `max_version` and initialize `next_txn_ts` when open.

### CODE REVIEW

**Comment by wangnengjie** [comment]
> LGTM, how about adding unit tests for `MemTables`?

https://github.com/tikv/agatedb/blob/master/src/memtable.rs#L94-L100

seems that memtable now assumes version of adding keys in monotone increase mode. Is that as expected?

**Comment by GanZiheng** [comment]
> > LGTM, how about adding unit tests for `MemTables`?
> 
> https://github.com/tikv/agatedb/blob/master/src/memtable.rs#L94-L100
> 
> seems that memtable now assumes version of adding keys in monotone increase mode. Is that as expected?

I think that is expected. Because we will call `memtable::MemTable::put` when writing updates to the LSM tree, and when using transaction with `write_ch_lock`, it is guaranteed that the timestamp of the updates pushed to the write channel is monotonically increasing.

But I think we may add comment for this or also update the `max_version` by comparing.

**Comment by wangnengjie** [comment]
> > > LGTM, how about adding unit tests for `MemTables`?
> > 
> > https://github.com/tikv/agatedb/blob/master/src/memtable.rs#L94-L100
> > 
> > seems that memtable now assumes version of adding keys in monotone increase mode. Is that as expected?
> 
> I think that is expected. Because we will call `memtable::MemTable::put` when writing updates to the LSM tree, and when using transaction with `write_ch_lock`, it is guaranteed that the timestamp of the updates pushed to the write channel is monotonically increasing.
> 
> But I think we may add comment for this or also update the `max_version` by comparing.

For managed_db mode, commit_ts is set by user. It might not be guaranteed?

**Comment by GanZiheng** [comment]
> > > > LGTM, how about adding unit tests for `MemTables`?
> > > 
> > > 
> > > https://github.com/tikv/agatedb/blob/master/src/memtable.rs#L94-L100
> > > seems that memtable now assumes version of adding keys in monotone increase mode. Is that as expected?
> > 
> > 
> > I think that is expected. Because we will call `memtable::MemTable::put` when writing updates to the LSM tree, and when using transaction with `write_ch_lock`, it is guaranteed that the timestamp of the updates pushed to the write channel is monotonically increasing.
> > But I think we may add comment for this or also update the `max_version` by comparing.
> 
> For managed_db mode, commit_ts is set by user. It might not be guaranteed?

Yes, so I think it would be better to update by comparing.

**Comment by wangnengjie** [comment]
Done

File: src/memtable.rs [hunk 1]
```diff
@@ -270,4 +294,54 @@ mod tests {
...
     }
+
+    #[test]
+    fn max_version() {
+        let skl = Skiplist::with_capacity(make_comparator(), 4 * 1024 * 1024, true);
+        let mem = MemTable::new(0, skl, None, AgateOptions::default());
+        for i in 200..300 {
+            let mut v = BytesMut::from(i.to_string().as_bytes());
+            append_ts(&mut v, i);
+            let v = v.freeze();
```
Comment (GanZiheng) [inline src/memtable.rs:305]: We could replace this with `key_with_ts`.

**Review by GanZiheng — APPROVED** [review]
Rest LGTM.

**Comment by GanZiheng** [comment]
PTAL @Connor1996.

**Review by GanZiheng — APPROVED** [review]

**Review by skyzh — APPROVED** [review]

**Comment by BusyJay** [comment]
It will be helpful if the description of the PR or a linked issue can explain what is `max_version`.

**Comment by wangnengjie** [comment]
> It will be helpful if the description of the PR or a linked issue can explain what is `max_version`.

Done. If somewhere is still puzzling, plz point out.

**Pull request merged**